File: src/0652-Find-Duplicate-Subtrees/0652.js
var findDuplicateSubtrees = function(root) {
    let res = [], m = {};
    let dfs = function(root) {
        if (!root) return "";
        let s = root.val + "," + dfs(root.left) + "," + dfs(root.right);
        m[s] = (m[s] || 0) + 1;
        if (m[s] == 2) res.push(root);
        return s;
    }
    dfs(root);
    return res;
};

File: src/0652-Find-Duplicate-Subtrees/0652.py
class Solution:
    def findDuplicateSubtrees(self, root: 'TreeNode') -> 'List[TreeNode]':
        d = collections.defaultdict(list)
        def dfs(root):
            if root:
                res = (root.val, dfs(root.left), dfs(root.right))
                d[res].append(root)
                return res
            return "None"
        dfs(root)
        return [v[0] for k, v in d.items() if len(v) > 1]